crypto/conf/conf_mod.c

```diff
@@ -161,16 +161,24 @@ int CONF_modules_load(const CONF *cnf, const char *appname,
 
     for (i = 0; i < sk_CONF_VALUE_num(values); i++) {
         vl = sk_CONF_VALUE_value(values, i);
-        ERR_set_mark();
+        /*
+         * we have to condition our mark setting here
+         * because parsing failures in the provider config
+         * aren't considered fatal and return success
```

Comment:
> Its meant to indicate that, because the config parser always returns success,

My understanding is that this behaviour is supposed to be controlled by the "conf_diagnostrics" line. So if "conf_diagnostics" is 1 then you are supposed to get failures, and not otherwise. So I'm not sure why we would make the behaviour different for this type of error.

Reply:
right, I think that diagnostic setting was broken.
 Prior to this change:
1) ERR_set_mark was called unilaterally at the start of the for loop in CONF_modules_load
2) module_run was called, which calls through module_init->...provider_conf_load->provider_conf_activate->provider_conf_params, the last of which detects a recursive section, and raises an error
3) provider_conf_load, even if an error is raised from provider_conf_params returns 1 because we consider failure to load a provider as non-fatal.  This bubbles back up to the module_run call in CONF_modules_load, meaning the only way to detect a recursive error outside of the return code values is the error stack
4)back in CONF_modules_load, the return code from module_run is 1 (success), and so the conditional (if ret <= 0) was not taken, and in its stead ERR_pop_to_mark was called, clearing the error stack back to where the mark was set at the beginning of the for loop, eliminating the error placed there by the recursion detection, preventing the application from seeing it.

I thought the best fix was to make the mark settings/clearning in CONF_modules_load conditional on the CONF_MFLAGS_IGNORE_ERRORS flag being set, so that in the test we can see the error stack produced by the provider config parsing.

Unless you see another way to propagate the error stack, or feel that provider error detection can be fatal, I'm not sure how we can test this condition

Reply:
In my mind there is a difference between a fundamental error in the config file itself, vs a valid config file that just fails to activate providers (e.g. because the .so file for the provider isn't present).

We could make an argument about whether it was the right choice to treat failure to load a provider as an error or not - but we're probably stuck with that choice (at least for now).

However, the recursive config file seems like a wholly different type of error to me. This is actually a problem in the config file itself, and can quite reasonably be treated as fatal. So, my preferred option here would be to cause provider_conf_load to return failure if a recursive config file is detected and 1 otherwise.

Reply:
Ok, I've:
* reverted the conf ERR pop changes
* added code in the provider conf parsing to update return codes to return a tristate, with 1 = success, 0 = non-fatal errors, <0 = fatal errors.  That is converted in provider_conf_load to a simple 1/0 return, where 0 is only returned on fatal errors
* Adjusted the test to ensure that we get a zero return code from OSSL_LIB_CTX_load_config